File: NativeChart/ios/Charts/Bar/NativeBarChart.swift
//
//  NativeBarChart.swift
//  NativeChart
//
//  Created by Yiheng Quan on 12/2/20.
//

import Charts

class NativeBarChart : BarChartView, OptimisedBarLineChart {
    @objc var chartData: [Double] = []
    @objc var legendLabel: String = ""
    // You will need one for each bar
    @objc var xAxisLabels: [String] = []
    @objc var darkMode: Bool = false
    @objc var themeColor: String = "#FFFFFF"
    
    override func didSetProps(_ changedProps: [String]!) {
        super.didSetProps(changedProps)
        self.updateChart()
    }
    
    init() {
        super.init(frame: CGRect())
        setupChart(chart: self)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func updateChart() {
        let textColour = darkMode ? UIColor.white : UIColor.black
        
        // Map double to ChartDataEntry
        let formattedData = self.chartData.enumerated().map({i, element in BarChartDataEntry(x: Double(i), y: element)})
        
        let chartDataSet = BarChartDataSet(entries: formattedData, label: legendLabel)
        // Use only theme colour
        chartDataSet.colors = [UIColor(hex: themeColor)!]
        chartDataSet.valueFont = UIFont.systemFont(ofSize: 12)
        chartDataSet.valueColors = [textColour]
        let chartData = BarChartData(dataSets: [chartDataSet])
        
        // Remove fractions
        let format = NumberFormatter()
        format.minimumFractionDigits = 0
        chartData.setValueFormatter(DefaultValueFormatter(formatter: format))
        
        // Update xAxis label
        self.xAxis.valueFormatter = IndexAxisValueFormatter(values: xAxisLabels)
        self.xAxis.setLabelCount(self.xAxisLabels.count, force: false)
        
        // Update chart data
        self.data = chartData
        
        // Update theme colour
        self.xAxis.labelTextColor = textColour
        self.leftAxis.labelTextColor = textColour
        self.legend.textColor = textColour
    }
    
    func setupChart(chart: BarLineChartViewBase) {
        // Text related
        chart.noDataText = "No Information are provided"
        chart.chartDescription.text = ""
        
        // Disable zoom and interaction
        chart.highlightPerTapEnabled = false
        chart.highlightPerDragEnabled = false
        chart.setScaleEnabled(false)

        // Custom style for xAxia
        chart.xAxis.labelPosition = .bottom
        chart.xAxis.drawGridLinesEnabled = false
        
        // Custom style for rightAxis
        chart.rightAxis.drawLabelsEnabled = false
        chart.rightAxis.drawGridLinesEnabled = false
        chart.rightAxis.drawAxisLineEnabled = false
        
        // Custom style for leftAxis
        chart.leftAxis.drawLabelsEnabled = false
        chart.leftAxis.drawGridLinesEnabled = false
        chart.leftAxis.drawAxisLineEnabled = false
        // Don't need to show the legend for this
        chart.legend.enabled = false
    }
}

File: NativeChart/android/src/main/java/org/github/henryquan/nativechart/bar/NativeBarChart.kt
package org.github.henryquan.nativechart.bar

import android.graphics.Color
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.components.LimitLine
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.*
import com.github.mikephil.charting.formatter.DefaultAxisValueFormatter
import com.github.mikephil.charting.formatter.DefaultValueFormatter
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter
import com.github.mikephil.charting.formatter.ValueFormatter
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet

class NativeBarChart(c: ThemedReactContext) : BarChart(c) {
    var chartData: ArrayList<Float> = arrayListOf()
    var xAxisLabels: ArrayList<String> = arrayListOf()
    var legendLabel: String = ""
    var darkMode: Boolean = false
    var themeColor: Int = Color.WHITE

    init {
        this.setupChart()
    }

    /**
     * Update chart
     */
    fun updateChart() {
        // Get theme colour
        val textColour = if (darkMode) Color.WHITE else Color.BLACK

        // Map float to entry
        val formattedData = this.chartData.mapIndexed { index, element -> BarEntry(index.toFloat(), element)}

        // Setup data and colour to chart
        val chartDataSet = BarDataSet(formattedData, this.legendLabel)
        chartDataSet.color = this.themeColor
        chartDataSet.valueTextSize = 12F
        chartDataSet.setValueTextColors(mutableListOf(textColour))
        chartDataSet.setDrawValues(false)



        // Set labels
        this.xAxis.valueFormatter = IndexAxisValueFormatter(this.xAxisLabels)
        this.xAxis.setLabelCount(this.xAxisLabels.size, false)

        // Update chart data
        val chartDataFormatted = BarData(mutableListOf<IBarDataSet>(chartDataSet))
        // Remove fractions
//        chartDataFormatted.setValueFormatter(DefaultAxisValueFormatter())

        this.data = chartDataFormatted

        // Update theme colour
        this.axisLeft.textColor = textColour
        this.xAxis.textColor = textColour
    }

    /**
     * Setup the chart
     */
    private fun setupChart() {
        // Text related
        this.setNoDataText("...")
        this.description.text = ""

        // Disable zoom and interaction
        this.isHighlightPerTapEnabled = false
        this.isHighlightPerDragEnabled = false
        this.setScaleEnabled(false)

        // Custom style for xAxia
        this.xAxis.position = XAxis.XAxisPosition.BOTTOM
        this.xAxis.setDrawGridLines(false)

        // Custom style for axisRight
        this.axisLeft.setDrawGridLines(false)
        this.axisLeft.setDrawLabels(false)
        this.axisLeft.setDrawAxisLine(false)

        // Custom style for axisLeft
        this.axisRight.setDrawGridLines(false)
        this.axisRight.setDrawLabels(false)
        this.axisRight.setDrawAxisLine(false)

        // Don't draw leend for this
        this.legend.isEnabled = false
    }
}